### PR TITLE
SLING-11061 URIProvider for resources backed by JCR BinaryDownload

### PR DESCRIPTION
**Pull request by kwin**
properties

### CODE REVIEW

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Provides URIs for direct binary read-access based on the Jackrabbit API {@link BinaryDownload}.
+ * 
+ * @see <a href="https://jackrabbit.apache.org/oak/docs/features/direct-binary-access.html">Oak Direct Binary Access</a>
+ *
+ */
+@Component(service = URIProvider.class, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@Designate(ocd = BinaryDownloadUriProvider.Configuration.class)
+public class BinaryDownloadUriProvider implements URIProvider {
```
Comment (ieb) [inline src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:50]: IIRC the URIProvider interface was exported with the intention that it would be implemented in other bundles ensuring that anyone could adapt how binaries are downloaded to their needs. There was a lot of discussion at the time it was introduced. 

I think it would be better to put this class in its own bundle so that it can have a life of its own and does not bind this bundle too tightly to the APIs it depends on.   

For example. The BinaryDownloadOptions class is relatively new and has gone through a few changes. I would expect it to go through more as more edge cases are discovered from real usage. I am expecting it to change in the future.

On ballance, I think this class is better in its own bundle. Is that possible ?

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java [hunk 2]
```diff
@@ -227,6 +212,31 @@ private InputStream getInputStream() {
         return null;
     }
 
+    static @NotNull Property getPrimaryProperty(@NotNull Node node) throws RepositoryException {
```
Comment (ieb) [inline src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:215]: This code is cleaner than before patch, but for the purposes of introducing a URIProvider implementation I dont think its is necessary, unless I missed something.

I would recommend that these changes are done in a separate PR if they are considered to have benefit worthy of making the change.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java [hunk 3]
```diff
@@ -181,34 +185,15 @@ private InputStream getInputStream() {
...
-                        // we don't actually care, but log for completeness
-                        LOGGER.debug("getInputStream: No primary items for {}", toString(), infe);
-                        data = null;
-                    }
+                try {
+                    data = getPrimaryProperty(node);
+                } catch (ItemNotFoundException infe) {
+                    // we don't actually care, but log for completeness
+                    LOGGER.debug("getInputStream: No primary items for {}", toString(), infe);
+                    data = null;
```
Comment (ieb) [inline src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:194]: This makes sure that `data` can be passed as null to all URIProvider implementations. 

Does the URIProvider API allow nulls in data. I suspect that other implementations will throw NPE after this change ?

**Review by ieb — CHANGES_REQUESTED** [review]
There is a potential new NPE in other URIProvider implementations that previously never got data as NULL. That needs to be be prevented. Hence requesting a change.

Also when URIProvider was introduced the intention was to implement it in bundles outside of this bundle so deployers could choose how they served binaries. Hence recommending a change.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:50]: I am not sure I follow this argument. I placed it in this bundle, because that already depends on Jackrabbit API and one method is shared between this and the `JcrNodeResource` (for getting the primary property). Anyone can easily override that by registering a service with a higher rank, also this service is only becoming active if it has been configured. 

> I am expecting it to change in the future.

The Jackrabbit API is stable, evolvements in the future will most probably be backwards compatible. Also given that this API has been introduced in July 2018 (https://issues.apache.org/jira/browse/JCR-4335) and never been changed since than, I don't see it being too volatile...

What do you want to gain with separating bundles? Just different release cycles or is there anything else I don't see?

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java (hunk 2 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:215]: This has been refactored because this method is used from the new URIProvider.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:194]: This hasn't changed in this PR, even before `data` could have been null (see removed line 209)

**Comment by kwin** [comment]
@rombert Any idea why this build fails on Jenkins? To me it looks like not all dependencies are downloaded...

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:50]: We can easily separate that later to a dedicated bundle (as this class is not exposed to other bundles) if we e.g. figure out that we need to depend on a newer Jackrabbit API that we don't feel comfortable adjust the requirements to in this bundle. But raising the requirements of this bundle now to Oak 1.10 is IMHO not a problem as this release is now three years old.

File: src/test/java/org/apache/sling/jcr/resource/internal/JcrResourceListenerTest.java [hunk 4]
```diff
@@ -269,7 +269,7 @@ public boolean matches(String path) {
                 session.save();
             }
             System.out.println("Events = " + events);
-            assertEquals("Received: " + events, 7, events.size());
+            assertEquals("Received: " + events, 6, events.size());
```
Comment (cziegeler) [inline src/test/java/org/apache/sling/jcr/resource/internal/JcrResourceListenerTest.java:272]: Why does this unrelated test needs to be adjusted?

Comment (kwin) [reply src/test/java/org/apache/sling/jcr/resource/internal/JcrResourceListenerTest.java:272]: See https://issues.apache.org/jira/browse/SLING-11063, therefore also in a dedicated commit https://github.com/apache/sling-org-apache-sling-jcr-resource/pull/19/commits/1cec8179d182cef5471fda8f1c0bff091347b3fd

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:49]: For every consumer (like AEM ;-)) already having another, potentially conflicting URIProvider in place, this won't be an issue, as this service is inactive by default!

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java (hunk 3 above)
Comment (ieb) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:194]: looking at it again, I think you are right.
The logic doesn't really make sense. Why does it spend time getting `data` when its only needed if convertToPublicURI returns null.

If there are no URIProviders that operation is a noop, but if there are URIProviders then they should be returning a uri.

No change requested here any more.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java (hunk 2 above)
Comment (ieb) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:215]: I think that is wrong, but its not for me to say.

**Review by ieb — COMMENTED** [review]
not requesting changes.
I don't feel I have been active in Sling enough recently to do approve or not a PR, please rely on others.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java:194]: > Why does it spend time getting data when its only needed if convertToPublicURI returns null.

data is needed or both cases!

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        if (!isRelevantScopeAndOperation(scope, operation)) {
+            throw new IllegalArgumentException("This provider does only provide URIs for read operations in scope public or external, but not for scope " + scope + " and operation " + operation);
+        }
+        Node node = resource.adaptTo(Node.class);
+        if (node == null) {
+            throw new IllegalArgumentException("This provider only provide URIs for node-based resources");
+        }
+        try {
+            // get main property (probably containing binary data)
+            Property primaryProperty = JcrNodeResource.getPrimaryProperty(node);
```
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:88]: This is a specific implementation for JCR backed resources only. I don't see a reason, why this should not reuse code from JcrNodeResource.

File: src/test/java/org/apache/sling/jcr/resource/internal/JcrResourceListenerTest.java (hunk 4 above)
Comment (cziegeler) [reply src/test/java/org/apache/sling/jcr/resource/internal/JcrResourceListenerTest.java:272]: Ok, thanks - makes sense

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java:88]: Its in the same bundle, so it is ok - I guess it would be nicer to move this to an util class to have clearer boundaries

[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/BinaryDownloadUriProvider.java]: Done in https://github.com/apache/sling-org-apache-sling-jcr-resource/pull/19/commits/802d27ea565610f021fae7173c49bbc5c04f9fbc.

**Review by cziegeler — APPROVED** [review]
Lgtm, thanks @kwin  for all the changes
